text/differential-layer-format.md

```diff
@@ -0,0 +1,143 @@
+Differential Layer
+==================
+
+## Current status
+
+Currently, there are two layer formats: Delta and Image. 
+
+Delta layers contain a sorted, indexed list of all changes to pages in a
+certain segment and a range of LSNs, whereas Image layers contain a snapshot of
+the pages at a certain LSN.
+
+The Differential Layer format tries to improve and replace the Delta layer format.
+
+## Problems with Delta layer; and improvements
+
+### Page Version Branches
+As mentioned in "Storage and file formats; Structures", storing page versions
+in branches is an easy optimization which makes finding the right data easy.
+This does mean that we will have one full page image in potential overhead for
+each modified page in the layer, but that is not too bad (with Delta layers
+this is one FPW per page in the segment, per delta layer, or we'd have
+cross-file dependencies).
+
+As there may be multiple kinds of access patterns to pages, it is beneficial to
+store some metadata in the BranchPtr which discerns the kind of branch that
+pointer points to. This allows for compacter storage of the actual branches,
+at the cost of some extra data in the
+
+Not all branches will describe multiple page versions: It is indeed possible
+that a page will be updated only once during a checkpoint.
+
+In that case, it
+is quite wasteful to store the metadata of '0 page versions'
+
+- FPI, 0 revisions
+- will-init WAL record, 0 revisions
+- FPI-initiated branch, N WAL records depending on that FPI
+- WAL-record initiated branch, N WAL records depending on that WAL record
+- ...
+
+In the physical format, because the LSNs of the WAL records are sorted and stored
+in increasing order, we could apply compression to this array to reduce the
+storage overhead of the full branch structure.
+
+Additionally, because the length of the WAL record is stored in it's header, we
+don't actually need to store the length of the WAL record in the branch structure;
+we can read it directly from the header, saving 4 bytes per WAL record.
+
+With LSN array compression and the emission of the WAL record length also being
+stored outside the WAL record, we can reduce the storage overhead of non-FPI WAL
+records from 28 bytes per record down to less than 8 bytes per record, with an
+expected value of ~ 4 bytes (depending on how often a page is modified; more
+frequent / less noizy WAL record generation = lower overhead).
+
```

Comment:
I partly agree with your arithmetic. But...
Right now delta_layer is storing 4 byte length + serialized PageVersion.
PageVersion itself is itself nested enum, with 8 bytes used just for enum tags (this seems to be the first optimization target, because one byte is enough to encode it). Size of Postgres XLogRecord header is 24 bytes.
Length is really stored inside XLogRecord. Typical size of record (i.e. update of integer field is 52 bytes - just check WAL log for pgbench).

So what we have. Right now we are definitely wasting 7 bytes just because of serialization, we can remove extra 4 bytes with record length. It will complicate traversal (to be able to deserialize data we first need to extract it and to know how much data we need to read, we need first to extract length). But in principle it is possible.
Using delta encoding we can encode LSN using just few bits. So we can throw away 20 bytes.
So instead of 12+52=74 bytes we will store 54 bytes. So we save about 30%. It is certainly  noticeable, but ... do not radically change storage and memory demands.

Reply:
am I looking at the wrong branch? [DeltaLayerInner](https://github.com/zenithdb/zenith/blob/main/pageserver/src/layered_repository/delta_layer.rs#L148) most definately uses an VecMap<(SegmentBlk, Lsn), BlobRange> to store the references to versions. `SegmentBlk (4) + Lsn (8) + BlobRange (8 + 8) = 28`. This is not even counting the overheads of the PageVersion storage itself.

> It is certainly noticeable, but ... do not radically change storage and memory demands.
The main part is that this optimization of Branches is targetted to storage; as the data of the Branch itself will only reside in memory whenever the page is actually being requested.

Sequentially stored Branches allow us to store the location of required data much more concisely; decreasing memory usage significantly.

Reply:
I looked mostly at in-memory layer layout.
I agree with you, 28 bytes is really too much overhead for in-memory part of delta layer.
Using RLE (run-length-encoding) and delta encoding we can really reduce 28 bytes to ~4 bytes.